tortoise/tortoise.go

```diff
@@ -1110,3 +1119,35 @@ func (t *turtle) ballotFilterForHealing(candidateLayerID types.LayerID, logger l
 		return lid.Uint32() > t.BadBeaconVoteDelayLayers && lid.Sub(t.BadBeaconVoteDelayLayers).After(candidateLayerID)
 	}
 }
+
+func getEpochWeight(atxdb atxDataProvider, epochWeight map[types.EpochID]uint64, eid types.EpochID) (uint64, error) {
+	weight, exist := epochWeight[eid]
+	if exist {
+		return weight, nil
+	}
+	weight, _, err := atxdb.GetEpochWeight(eid)
+	if err != nil {
+		return 0, fmt.Errorf("epoch weight %s: %w", eid, err)
+	}
+	epochWeight[eid] = weight
+	return weight, nil
+}
+
+// computeExpectedVoteWeight computes the expected weight for a block in the target layer, from summing up votes from ballots up to and including last layer.
+func computeExpectedVoteWeight(atxdb atxDataProvider, epochWeight map[types.EpochID]uint64, target, last types.LayerID) (*big.Float, error) {
+	var (
+		total  = new(big.Float)
+		layers = uint64(types.GetLayersPerEpoch())
+	)
+	// TODO(dshulyak) this can be improved by computing weight for epoch at once
+	for lid := target.Add(1); !lid.After(last); lid = lid.Add(1) {
+		weight, err := getEpochWeight(atxdb, epochWeight, lid.GetEpoch())
+		if err != nil {
+			return nil, err
+		}
+		layerWeight := new(big.Float).SetUint64(weight)
+		layerWeight.Quo(layerWeight, new(big.Float).SetUint64(layers))
```

Comment:
i read Tal's comment after i commented. i think he was looking for max weight in that epoch to calculate the max possible weight here. but i think this is also good enough.

Reply:
can you clarify? in my understanding this is exactly what he meant, there is even similar tests:

> To give a concrete example, suppose epochs are exactly 100 layers each, block X is in layer 80 and we're currently in layer 210. Also, suppose the total ATX weight for the first epoch is 30, for the second epoch 40 and for the third epoch 50.  Then in the first epoch, there are only 20 (out of 100) layers that can vote about block X, so we count the maximum weight in that epoch as 30*20/100 = 6. For the second epoch, all layers are eligible to vote, so the total voting weight is 40. In the third epoch, there are only 10 layers so far, so the expected voting weight is 50*10/100 = 5. Thus, for the purpose of the global threshold,  the total voting weight is 6+40+5=51.

the only difference is that i am computing full layer weight at once, instead it will added incrementally

Reply:
you are right. i misread Tal's comment. i thought he meant to find the max weight in that epoch and extrapolate. my bad.